modules/apps/layout/layout-content-page-editor-web-test/src/testIntegration/java/com/liferay/layout/content/page/editor/web/internal/portlet/action/test/DuplicateSegmentsExperienceMVCActionCommandTest.java

```diff
@@ -0,0 +1,219 @@
+ */
+@RunWith(Arquillian.class)
+public class DuplicateSegmentsExperienceMVCActionCommandTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+
+		_layout = _addLayout();
+
+		ServiceContextThreadLocal.pushServiceContext(new ServiceContext());
+	}
+
+	@After
+	public void tearDown() {
+		ServiceContextThreadLocal.popServiceContext();
+	}
+
+	@Test
+	public void testDuplicateSegmentsExperience() throws Exception {
+		SegmentsExperience segmentsExperience = _addSegmentsExperience();
+
+		MockLiferayPortletActionRequest mockLiferayPortletActionRequest =
+			_getMockLiferayPortletActionRequest();
+
+		mockLiferayPortletActionRequest.addParameter(
+			"segmentsExperienceId",
+			String.valueOf(segmentsExperience.getSegmentsExperienceId()));
+
+		JSONObject jsonObject = ReflectionTestUtil.invoke(
+			_duplicateSegmentsExperienceMVCActionCommand,
+			"doTransactionalCommand",
+			new Class<?>[] {ActionRequest.class, ActionResponse.class},
+			mockLiferayPortletActionRequest,
+			new MockLiferayPortletActionResponse());
+
+		JSONObject segmentsExperienceJSONObject = jsonObject.getJSONObject(
+			"segmentsExperience");
+
+		Assert.assertEquals(
+			segmentsExperience.isActive(),
+			segmentsExperienceJSONObject.getBoolean("active"));
+		Assert.assertEquals(
+			"Experience (Copy)",
+			segmentsExperienceJSONObject.getString("name"));
+		Assert.assertEquals(
+			segmentsExperience.getPriority() + 1,
+			segmentsExperienceJSONObject.getInt("priority"));
+		Assert.assertEquals(
+			segmentsExperience.getSegmentsEntryId(),
+			segmentsExperienceJSONObject.getLong("segmentsEntryId"));
+		Assert.assertTrue(
+			segmentsExperienceJSONObject.getLong("segmentsExperienceId") > 0);
+	}
+
+	private Layout _addLayout() throws Exception {
+		ServiceContext serviceContext =
+			ServiceContextTestUtil.getServiceContext(
+				TestPropsValues.getGroupId(), TestPropsValues.getUserId());
+
+		return _layoutLocalService.addLayout(
+			TestPropsValues.getUserId(), _group.getGroupId(), false,
+			LayoutConstants.DEFAULT_PARENT_LAYOUT_ID,
+			RandomTestUtil.randomString(), RandomTestUtil.randomString(),
+			StringPool.BLANK, LayoutConstants.TYPE_CONTENT, false,
+			StringPool.BLANK, serviceContext);
+	}
+
+	private SegmentsExperience _addSegmentsExperience() throws Exception {
+		MockLiferayPortletActionRequest mockLiferayPortletActionRequest =
+			_getMockLiferayPortletActionRequest();
+
+		mockLiferayPortletActionRequest.addParameter("name", "Experience");
+
+		SegmentsEntry segmentsEntry = SegmentsTestUtil.addSegmentsEntry(
+			_group.getGroupId());
+
+		mockLiferayPortletActionRequest.addParameter(
+			"segmentsEntryId",
+			String.valueOf(segmentsEntry.getSegmentsEntryId()));
+
+		JSONObject jsonObject = ReflectionTestUtil.invoke(
```

Comment:
Maybe we could call the service directly?

Reply:
you meant, without creating a new method in the service layer? well, of course we could, but we're going to need it to recalculate the right position, don't you think?

Reply:
I will reevaluate it, once we have this requirement (sinde we can create, for the test, the experience in the place we want, and it is safer than call a MVCActionCommand).

Reply:
But what I mean is the comment is that the method _addSegmentsExperience, calls the method addSegmentsExperience, in _addSegmentsExperienceMVCActionCommand. I think it is usually a not good idea, to depend on this complex code. 

In the future your code will fail when the code of the _addSegmentsExperienceMVCActionCommand is changed, and this is something that you probably want to avoid.

Usually we prefer to depends on the service layer for the creation of the context where the tests should be executed.

Reply:
I understand now, I'm going to try it. As you see there's a lot of logic in that MVC Command apart from creating the segments experience in internal methods. I hope I can achieve to recreate it in the test